### PR TITLE
Feature/integration custom transform

### PR DESCRIPTION
**Pull request by jan-schutte**

### CODE REVIEW

File: src/Nox.Integration/ServiceExtension.cs
Comment (rochar) [inline src/Nox.Integration/ServiceExtension.cs:1]: @jan-schutte  a small refactor for consistent
Can you add this file to Extenison folder an call it ServiceCollectionExtensions

File: src/Nox.Integration/ServiceExtension.cs [hunk 1]
```diff
@@ -16,12 +16,19 @@ public static IServiceCollection AddNoxIntegrations(this IServiceCollection serv
             services.TryAddSingleton<INoxIntegrationContext>(p =>
             {
                 var logger = p.GetRequiredService<ILogger<INoxIntegrationContext>>();
-                var context = new NoxIntegrationContext(logger, solution);
+                var handlers = p.GetServices<INoxCustomTransformHandler>();
+                var context = new NoxIntegrationContext(logger, solution, handlers);
```
Comment (rochar) [inline src/Nox.Integration/ServiceExtension.cs]: NoxIntegrationContext can request all dependencies in the constructor
NoxSolution, ILogger<T> and INoxCustomTransformHandler are all registered in the container

File: src/Nox.Integration/ServiceExtension.cs [hunk 2]
```diff
@@ -16,12 +16,19 @@ public static IServiceCollection AddNoxIntegrations(this IServiceCollection serv
...
                 context.Initialize();
```
Comment (rochar) [inline src/Nox.Integration/ServiceExtension.cs]: Can  the Initialized be called in the constructor, and so only initialize if the NoxIntegrationContext service is requested?

File: src/Nox.Integration/Services/NoxIntegration.cs
Comment (rochar) [inline src/Nox.Integration/Services/NoxIntegration.cs:1]: NoxIntegration can be internal and sealed? and the interface also? its not to be consumed outside this component?

File: src/Nox.Integration/Services/NoxIntegrationContext.cs [hunk 3]
```diff
@@ -25,16 +28,22 @@ public void Initialize()
...
     
     public async Task<bool> ExecuteIntegrationAsync(string name)
     {
-        var integration = _integrations.Single(i => i.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
-        var result = await integration.ExecuteAsync();
-        return result;
+        try
+        {
+            var integration = _integrations.Single(i => i.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
+            var result = await integration.ExecuteAsync(_handlers);
```
Comment (rochar) [inline src/Nox.Integration/Services/NoxIntegrationContext.cs]: on my perspective the integration should already receive the proper handler, according to the integration name and if not found raise the NoxIntegrationException directly

File: src/Nox.Integration/Services/NoxIntegration.cs [hunk 4]
```diff
@@ -46,33 +50,60 @@ public NoxIntegration(ILogger logger, Solution.Integration definition)
...
+
+        IDataFlowSource<ExpandoObject>? transformSource = null;
+        
+        if (TransformType == IntegrationTransformType.CustomTransform)
+        {
+            if (handlers == null) throw new NoxIntegrationException("Cannot execute custom transform, no handlers registered.");
+            //Find the handler
+            try
+            {
+                var handler = handlers.Single(h => h.IntegrationName == Name);
```
Comment (rochar) [inline src/Nox.Integration/Services/NoxIntegration.cs]: IMO the handler can be selected on NoxContext like the  NoxIntegration is
NoxIntegration could only know ist proper handler

File: src/Nox.Integration/Services/NoxIntegrationContext.cs [hunk 5]
```diff
@@ -10,12 +11,14 @@ public class NoxIntegrationContext: INoxIntegrationContext
...
     private readonly NoxSolution _solution;
+    private readonly IEnumerable<INoxCustomTransformHandler>? _handlers;
     
-    public NoxIntegrationContext(ILogger<INoxIntegrationContext> logger, NoxSolution solution)
+    public NoxIntegrationContext(ILogger<INoxIntegrationContext> logger, NoxSolution solution, IEnumerable<INoxCustomTransformHandler>? handlers = null)
     {
         _logger = logger;
         _integrations = new List<INoxIntegration>();
         _solution = solution;
+        _handlers = handlers;
```
Comment (rochar) [inline src/Nox.Integration/Services/NoxIntegrationContext.cs]: here we could add the handlers to a Dictionary<string,INoxCustomTransformHandler> so on execution no need to search by name

File: src/Nox.Integration/Services/NoxIntegrationContext.cs [hunk 6]
```diff
@@ -25,16 +28,22 @@ public void Initialize()
             var instance = new NoxIntegration(_logger, integrationDefinition);
             instance.WithReceiveAdapter(integrationDefinition.Source, _solution.Infrastructure?.Dependencies?.DataConnections);
             instance.WithSendAdapter(integrationDefinition.Target, _solution.Infrastructure?.Dependencies?.DataConnections);
-            
             _integrations.Add(instance);
```
Comment (rochar) [inline src/Nox.Integration/Services/NoxIntegrationContext.cs]: _integrations could be a Dictionary<string,NoxIntegration> that way we would search by integration name in O(1), no need to do a O(n) every time

File: src/Nox.Integration/Services/NoxIntegrationContext.cs [hunk 7]
```diff
@@ -10,12 +11,14 @@ public class NoxIntegrationContext: INoxIntegrationContext
     private readonly ILogger _logger;
     private readonly List<INoxIntegration> _integrations;
     private readonly NoxSolution _solution;
+    private readonly IEnumerable<INoxCustomTransformHandler>? _handlers;
     
-    public NoxIntegrationContext(ILogger<INoxIntegrationContext> logger, NoxSolution solution)
+    public NoxIntegrationContext(ILogger<INoxIntegrationContext> logger, NoxSolution solution, IEnumerable<INoxCustomTransformHandler>? handlers = null)
```
Comment (rochar) [inline src/Nox.Integration/Services/NoxIntegrationContext.cs:16]: we can make this class internal and sealed

**Review by rochar — CHANGES_REQUESTED** [review]

File: src/Nox.Integration/ServiceExtension.cs (hunk 1 above)
Comment (jan-schutte) [reply src/Nox.Integration/ServiceExtension.cs]: Implemented

File: src/Nox.Integration/ServiceExtension.cs (hunk 2 above)
Comment (jan-schutte) [reply src/Nox.Integration/ServiceExtension.cs]: Implemented in constructor

File: src/Nox.Integration/Services/NoxIntegrationContext.cs (hunk 3 above)
Comment (jan-schutte) [reply src/Nox.Integration/Services/NoxIntegrationContext.cs]: Implemented

File: src/Nox.Integration/Services/NoxIntegrationContext.cs (hunk 5 above)
Comment (jan-schutte) [reply src/Nox.Integration/Services/NoxIntegrationContext.cs]: Implemented

File: src/Nox.Integration/Services/NoxIntegrationContext.cs (hunk 7 above)
Comment (jan-schutte) [reply src/Nox.Integration/Services/NoxIntegrationContext.cs:16]: Implemented

File: src/Nox.Integration/Services/NoxIntegrationContext.cs (hunk 6 above)
Comment (jan-schutte) [reply src/Nox.Integration/Services/NoxIntegrationContext.cs]: Implemented

File: src/Nox.Integration/ServiceExtension.cs
Comment (jan-schutte) [reply src/Nox.Integration/ServiceExtension.cs:1]: Implemented

**Pull request merged**